File: packages/nerva-js/src/tracing/json-log.ts
/**
 * JSON log tracer — structured trace output to file or writable stream.
 *
 * Writes one JSON line per trace callback (span start, span end, event,
 * complete). Useful for local debugging, log aggregation pipelines, and
 * audit trails.
 *
 * @module tracing/json-log
 */

import * as fs from "node:fs";
import * as path from "node:path";

import type { ExecContext, Event, Span } from "./index.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Trace event type for span start records. */
const SPAN_START_TYPE = "span_start";

/** Trace event type for span end records. */
const SPAN_END_TYPE = "span_end";

/** Trace event type for discrete event records. */
const EVENT_TYPE = "event";

/** Trace event type for request completion records. */
const COMPLETE_TYPE = "complete";

// ---------------------------------------------------------------------------
// Writable abstraction
// ---------------------------------------------------------------------------

/** Minimal writable interface — anything with a `write` method. */
export interface Writable {
  write(data: string): void;
}

// ---------------------------------------------------------------------------
// JsonLogTracer
// ---------------------------------------------------------------------------

/**
 * Writes trace events as JSON lines to a file or writable stream.
 *
 * Each span start/end and event produces one JSON line containing a
 * timestamp, event type, request/trace identifiers, and relevant data
 * fields from the span or event.
 */
export class JsonLogTracer {
  private readonly stream: Writable;
  private readonly ownsFile: boolean;

  /**
   * @param output - Where to write JSON lines. Accepts a file path (string)
   *   which will be opened in append mode, or any object with a `write(string)`
   *   method. Defaults to `process.stderr`.
   */
  constructor(output?: string | Writable) {
    const resolved = resolveOutput(output);
    this.stream = resolved.stream;
    this.ownsFile = resolved.ownsFile;
  }

  /**
   * Write a JSON line when a span begins.
   *
   * @param span - The span that just started.
   * @param ctx - Execution context the span belongs to.
   */
  onSpanStart(span: Span, ctx: ExecContext): void {
    const payload = buildSpanPayload(span);
    this.writeLine(SPAN_START_TYPE, ctx, payload);
  }

  /**
   * Write a JSON line when a span ends.
   *
   * Includes the span's duration in seconds when `endedAt` is available.
   *
   * @param span - The span that just ended.
   * @param ctx - Execution context the span belongs to.
   */
  onSpanEnd(span: Span, ctx: ExecContext): void {
    const payload = buildSpanPayload(span);
    if (span.endedAt !== null) {
      payload["duration_s"] = round6(span.endedAt - span.startedAt);
    }
    this.writeLine(SPAN_END_TYPE, ctx, payload);
  }

  /**
   * Write a JSON line when an event is recorded.
   *
   * @param event - The point-in-time event.
   * @param ctx - Execution context the event belongs to.
   */
  onEvent(event: Event, ctx: ExecContext): void {
    const payload = buildEventPayload(event);
    this.writeLine(EVENT_TYPE, ctx, payload);
  }

  /**
   * Write a JSON line when a request completes.
   *
   * Includes summary fields: total span count, event count, and
   * accumulated token usage.
   *
   * @param ctx - The completed execution context.
   */
  onComplete(ctx: ExecContext): void {
    const payload = buildCompletePayload(ctx);
    this.writeLine(COMPLETE_TYPE, ctx, payload);
  }

  /**
   * Close the underlying stream if this tracer owns it.
   *
   * Safe to call multiple times. Only closes file handles that were
   * opened by this tracer (not caller-provided streams or stderr).
   */
  close(): void {
    if (this.ownsFile && "end" in this.stream) {
      (this.stream as fs.WriteStream).end();
    }
  }

  // -- Private helpers ----------------------------------------------------

  /**
   * Serialize and write a single JSON line to the output stream.
   *
   * @param eventType - One of the trace event type constants.
   * @param ctx - Execution context for identity fields.
   * @param payload - Type-specific data to merge into the line.
   */
  private writeLine(
    eventType: string,
    ctx: ExecContext,
    payload: Record<string, unknown>,
  ): void {
    const line = buildBaseRecord(eventType, ctx);
    Object.assign(line, payload);
    this.stream.write(JSON.stringify(line) + "\n");
  }
}

// ---------------------------------------------------------------------------
// Output resolution
// ---------------------------------------------------------------------------

/** Result of resolving the output target. */
interface ResolvedOutput {
  readonly stream: Writable;
  readonly ownsFile: boolean;
}

/**
 * Turn the user-provided output target into a writable stream.
 *
 * @param output - File path, writable stream, or undefined for stderr.
 * @returns The resolved stream and whether this tracer owns it.
 */
function resolveOutput(output: string | Writable | undefined): ResolvedOutput {
  if (output === undefined) {
    return { stream: process.stderr, ownsFile: false };
  }

  if (typeof output === "string") {
    const resolved = path.resolve(output);
    const dir = path.dirname(resolved);
    fs.mkdirSync(dir, { recursive: true });
    const stream = fs.createWriteStream(resolved, { flags: "a", encoding: "utf-8" });
    return { stream, ownsFile: true };
  }

  return { stream: output, ownsFile: false };
}

// ---------------------------------------------------------------------------
// Pure helpers — build JSON-serialisable records
// ---------------------------------------------------------------------------

/**
 * Create the common envelope shared by every JSON line.
 *
 * @param eventType - The trace event type string.
 * @param ctx - Execution context to pull identifiers from.
 * @returns Record with type, timestamp, request_id, and trace_id.
 */
function buildBaseRecord(
  eventType: string,
  ctx: ExecContext,
): Record<string, unknown> {
  return {
    type: eventType,
    timestamp: Date.now() / 1000,
    request_id: ctx.requestId,
    trace_id: ctx.traceId,
  };
}

/**
 * Extract span fields into a JSON-ready record.
 *
 * @param span - The span to serialise.
 * @returns Record containing span identity, name, timing, and attributes.
 */
function buildSpanPayload(span: Span): Record<string, unknown> {
  return {
    span_id: span.spanId,
    span_name: span.name,
    parent_id: span.parentId,
    started_at: span.startedAt,
    ended_at: span.endedAt,
    attributes: { ...span.attributes },
  };
}

/**
 * Extract event fields into a JSON-ready record.
 *
 * @param event - The event to serialise.
 * @returns Record containing event name, timestamp, and attributes.
 */
function buildEventPayload(event: Event): Record<string, unknown> {
  return {
    event_name: event.name,
    event_timestamp: event.timestamp,
    attributes: { ...event.attributes },
  };
}

/**
 * Build summary data for the completion record.
 *
 * @param ctx - The completed execution context.
 * @returns Record with span/event counts and token usage breakdown.
 */
function buildCompletePayload(ctx: ExecContext): Record<string, unknown> {
  const elapsedSeconds = (Date.now() / 1000) - ctx.createdAt;
  return {
    span_count: ctx.spans.length,
    event_count: ctx.events.length,
    token_usage: {
      prompt_tokens: ctx.tokenUsage.promptTokens,
      completion_tokens: ctx.tokenUsage.completionTokens,
      total_tokens: ctx.tokenUsage.totalTokens,
      cost_usd: ctx.tokenUsage.costUsd,
    },
    elapsed_s: round6(elapsedSeconds),
  };
}

/**
 * Round a number to 6 decimal places.
 *
 * @param value - The number to round.
 * @returns The rounded number.
 */
function round6(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}

File: packages/nerva-py/nerva/tracing/json_log.py
"""JSON log tracer — structured trace output to file or stderr.

Writes one JSON line per trace callback (span start, span end, event,
complete). Useful for local debugging, log aggregation pipelines, and
audit trails.
"""

from __future__ import annotations

import json
import sys
import time
from dataclasses import asdict
from pathlib import Path
from typing import TextIO

from nerva.context import ExecContext, Event, Span


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

SPAN_START_TYPE = "span_start"
SPAN_END_TYPE = "span_end"
EVENT_TYPE = "event"
COMPLETE_TYPE = "complete"


class JsonLogTracer:
    """Writes trace events as JSON lines to a file or stream.

    Each span start/end and event produces one JSON line containing a
    timestamp, event type, request/trace identifiers, and relevant data
    fields from the span or event.

    Args:
        output: Where to write JSON lines. Accepts a file path (``str`` or
            ``Path``) which will be opened in append mode, or a writable
            ``TextIO`` stream. Defaults to ``sys.stderr``.
    """

    def __init__(self, output: str | Path | TextIO | None = None) -> None:
        self._owns_file = False
        self._stream = self._resolve_output(output)

    def on_span_start(self, span: Span, ctx: ExecContext) -> None:
        """Write a JSON line when a span begins.

        Args:
            span: The span that just started.
            ctx: Execution context the span belongs to.
        """
        payload = _build_span_payload(span)
        self._write_line(SPAN_START_TYPE, ctx, payload)

    def on_span_end(self, span: Span, ctx: ExecContext) -> None:
        """Write a JSON line when a span ends.

        Includes the span's duration in seconds.

        Args:
            span: The span that just ended.
            ctx: Execution context the span belongs to.
        """
        payload = _build_span_payload(span)
        if span.ended_at is not None:
            payload["duration_s"] = round(span.ended_at - span.started_at, 6)
        self._write_line(SPAN_END_TYPE, ctx, payload)

    def on_event(self, event: Event, ctx: ExecContext) -> None:
        """Write a JSON line when an event is recorded.

        Args:
            event: The point-in-time event.
            ctx: Execution context the event belongs to.
        """
        payload = _build_event_payload(event)
        self._write_line(EVENT_TYPE, ctx, payload)

    def on_complete(self, ctx: ExecContext) -> None:
        """Write a JSON line when a request completes.

        Includes summary fields: total span count, event count, and
        accumulated token usage.

        Args:
            ctx: The completed execution context.
        """
        payload = _build_complete_payload(ctx)
        self._write_line(COMPLETE_TYPE, ctx, payload)

    def close(self) -> None:
        """Close the underlying stream if this tracer owns it.

        Safe to call multiple times. Only closes file handles that were
        opened by this tracer (not caller-provided streams or stderr).
        """
        if self._owns_file and not self._stream.closed:
            self._stream.close()

    # -- Private helpers ----------------------------------------------------

    def _resolve_output(self, output: str | Path | TextIO | None) -> TextIO:
        """Turn the user-provided output target into a writable stream.

        Args:
            output: File path, stream, or ``None`` for stderr.

        Returns:
            A writable ``TextIO`` stream.
        """
        if output is None:
            return sys.stderr

        if isinstance(output, (str, Path)):
            path = Path(output)
            path.parent.mkdir(parents=True, exist_ok=True)
            self._owns_file = True
            return open(path, "a", encoding="utf-8")  # noqa: SIM115

        return output

    def _write_line(
        self,
        event_type: str,
        ctx: ExecContext,
        payload: dict[str, object],
    ) -> None:
        """Serialize and write a single JSON line to the output stream.

        Args:
            event_type: One of the ``*_TYPE`` constants.
            ctx: Execution context for identity fields.
            payload: Type-specific data to merge into the line.
        """
        line = _build_base_record(event_type, ctx)
        line.update(payload)
        self._stream.write(json.dumps(line, default=str) + "\n")
        self._stream.flush()


# ---------------------------------------------------------------------------
# Pure helpers — build JSON-serialisable dicts
# ---------------------------------------------------------------------------


def _build_base_record(event_type: str, ctx: ExecContext) -> dict[str, object]:
    """Create the common envelope shared by every JSON line.

    Args:
        event_type: The trace event type string.
        ctx: Execution context to pull identifiers from.

    Returns:
        Dict with ``type``, ``timestamp``, ``request_id``, and ``trace_id``.
    """
    return {
        "type": event_type,
        "timestamp": time.time(),
        "request_id": ctx.request_id,
        "trace_id": ctx.trace_id,
    }


def _build_span_payload(span: Span) -> dict[str, object]:
    """Extract span fields into a JSON-ready dict.

    Args:
        span: The span to serialise.

    Returns:
        Dict containing span identity, name, timing, and attributes.
    """
    return {
        "span_id": span.span_id,
        "span_name": span.name,
        "parent_id": span.parent_id,
        "started_at": span.started_at,
        "ended_at": span.ended_at,
        "attributes": dict(span.attributes),
    }


def _build_event_payload(event: Event) -> dict[str, object]:
    """Extract event fields into a JSON-ready dict.

    Args:
        event: The event to serialise.

    Returns:
        Dict containing event name, timestamp, and attributes.
    """
    return {
        "event_name": event.name,
        "event_timestamp": event.timestamp,
        "attributes": dict(event.attributes),
    }


def _build_complete_payload(ctx: ExecContext) -> dict[str, object]:
    """Build summary data for the completion record.

    Args:
        ctx: The completed execution context.

    Returns:
        Dict with span/event counts and token usage breakdown.
    """
    return {
        "span_count": len(ctx.spans),
        "event_count": len(ctx.events),
        "token_usage": asdict(ctx.token_usage),
        "elapsed_s": round(ctx.elapsed_seconds(), 6),
    }
